### PR TITLE
Output a simplified yaml for model-config

### PR DESCRIPTION
**Pull request by SimonRichardson**
## Description of change

The following changes allow us to output a model-config yaml that can
also be consumed by model-config. The type is then isomorphic, as in we
don't lose any information through the usage of this command.

This doesn't break backward compatibility but introduces a new output
format: simple-yaml (better name?), that removes the superfluous
information that isn't required by the input for the command.

We could have attempted to parse the original yaml into the simplified
yaml, but we would have issues knowing if it really did conform to the
simplified yaml and would be very error-prone.

-----

The command is actually really simple if you supply `--all` or
`--format=simple-yaml` then it will output a nice neat version of the
yaml. All here, just redirects the previous yaml output to the simple
yaml output, killing two birds with one stone.

-----

The only thing to consider is that `agent-version` is not allowed to be
piped directly into model-config, so is required to be removed. This is
a damn shame and would have been better to have a flag on model-config,
something like --ignore-error.

To get around this, you can do:

```
juju model-config --all > output.yaml && \
    sed -i '/^agent\-version/d' output.yaml && \
    juju model-config ./output.yaml
```

#### Output then update

Output `model-config`, then update the `model-config`:

```sh
$ juju model-config --all > output.yaml && \
    sed -i '/^agent\-version/d' output.yaml && \
    juju model-config ./output.yaml
```

#### Output all

```sh
$ juju model-config --all         
agent-metadata-url: ""
agent-stream: released
agent-version: 2.8.1.1
apt-ftp-proxy: ""
apt-http-proxy: http://172.17.0.2:3142
apt-https-proxy: ""
apt-mirror: ""
apt-no-proxy: ""
automatically-retry-hooks: true
backup-dir: ""
cloudinit-userdata: |
  packages:
    - 'jq'
container-image-metadata-url: ""
container-image-stream: released
container-inherit-properties: ""
container-networking-method: local
default-series: bionic
default-space: ""
development: false
disable-network-management: false
egress-subnets: ""
enable-os-refresh-update: true
enable-os-upgrade: true
fan-config: ""
firewall-mode: instance
ftp-proxy: ""
http-proxy: ""
https-proxy: ""
ignore-machine-addresses: false
image-metadata-url: ""
image-stream: released
juju-ftp-proxy: ""
juju-http-proxy: ""
juju-https-proxy: ""
juju-no-proxy: 127.0.0.1,localhost,::1
logforward-enabled: false
logging-config: <root>=DEBUG
lxd-snap-channel: latest/stable
max-action-results-age: 336h
max-action-results-size: 5G
max-status-history-age: 336h
max-status-history-size: 5G
net-bond-reconfigure-delay: 17
no-proxy: 127.0.0.1,localhost,::1
provisioner-harvest-mode: destroyed
proxy-ssh: false
resource-tags: a=1 b=2
snap-http-proxy: ""
snap-https-proxy: ""
snap-store-assertions: ""
snap-store-proxy: ""
snap-store-proxy-url: ""
ssl-hostname-verification: true
storage-default-filesystem-source: lxd
test-mode: false
transmit-vendor-metrics: true
update-status-hook-interval: 5m
```

#### Output one

```sh
$ juju model-config --all cloudinit-userdata       
cloudinit-userdata: |
  packages:
    - 'jq'
```

## Bug reference

https://bugs.launchpad.net/juju/+bug/1879972

### CODE REVIEW

File: tests/main.sh [hunk 1]
```diff
@@ -162,7 +162,7 @@ while getopts "hH?vVtAs:a:x:rl:p:S:" opt; do
         export BOOTSTRAP_REUSE_LOCAL="${OPTARG}"
         export BOOTSTRAP_REUSE="true"
         CLOUD=$(juju show-controller "${OPTARG}" --format=json | jq -r ".[\"${OPTARG}\"] | .details | .cloud")
-        PROVIDER=$(juju clouds --client 2>/dev/null | grep "${CLOUD}" | awk '{print $4}')
+        PROVIDER=$(juju clouds --client 2>/dev/null | grep "${CLOUD}" | awk '{print $4}' | head -n 1)
```
Comment (SimonRichardson) [inline tests/main.sh:165]: Drive-by fix from @achilleasa recent changes, if you have multiple LXD clouds, this breaks.

File: cmd/juju/model/configcommand.go [hunk 2]
```diff
@@ -70,11 +71,12 @@ type configCommand struct {
...
-	keys       []string
-	reset      []string // Holds the keys to be reset until parsed.
-	resetKeys  []string // Holds the keys to be reset once parsed.
-	setOptions common.ConfigFlag
+	action           func(configCommandAPI, *cmd.Context) error // The action which we want to handle, set in cmd.Init.
+	keys             []string
+	reset            []string // Holds the keys to be reset until parsed.
+	resetKeys        []string // Holds the keys to be reset once parsed.
+	setOptions       common.ConfigFlag
+	outputSimpleYAML bool
```
Comment (manadart) [inline cmd/juju/model/configcommand.go:79]: `outputConsumableYAML` or something?

**Comment by SimonRichardson** [comment]
> We need to update the help text to reflect the new behaviour.
> 
> I found it weird that using `--all` with `--format json` or `--format yaml` effectively disregarded the `--all` flag. Instead of only intercepting the tabular format, I think we should be doing the conversion to map earlier and supplying that result to the formatters.

This is why I don't like `--all` flag. It doesn't do what you want to all the formats. So `json` and `yaml` output correctly the formats, it outputs the `cloudinit-userdata` in the right format. The problem occurs for `tabular` data and `yaml` data *only* when you've selected a key. For `tabular` data we should leave as is, as adding the output breaks the `tabular` format. It looks bad.

If we output the `cloudinit-userdata` in tabular format (
```sh
juju model-config --all
Attribute                  From        Value
agent-metadata-url         default     ""
agent-stream               default     released
agent-version              model       2.8.1.1
apt-ftp-proxy              default     ""
apt-http-proxy             controller  http://172.17.0.2:3142
apt-https-proxy            default     ""
apt-mirror                 default     ""
apt-no-proxy               default     ""
automatically-retry-hooks  default     true
backup-dir                 default     ""
cloudinit-userdata         model       |
  packages:
    - 'jq'
    - 'pastebinit'
  runcmd:
    - [ sh, -xc, "echo $(date) ': hello world!'" ]
    - [ sh, -c, echo "=========hello world'=========" ]
container-image-metadata-url       default  ""
container-image-stream             default  released
container-inherit-properties       default  ""
container-networking-method        model    local
default-series                     default  bionic
default-space                      default  ""
development                        default  false
disable-network-management         default  false
egress-subnets                     default  ""
enable-os-refresh-update           default  true
enable-os-upgrade                  default  true
fan-config                         default  ""
firewall-mode                      default  instance
ftp-proxy                          default  ""
http-proxy                         default  ""
https-proxy                        default  ""
ignore-machine-addresses           default  false
image-metadata-url                 default  ""
image-stream                       default  released
juju-ftp-proxy                     default  ""
juju-http-proxy                    default  ""
juju-https-proxy                   default  ""
juju-no-proxy                      default  127.0.0.1,localhost,::1
logforward-enabled                 default  false
logging-config                     model    <root>=DEBUG
lxd-snap-channel                   default  latest/stable
max-action-results-age             default  336h
max-action-results-size            default  5G
max-status-history-age             default  336h
max-status-history-size            default  5G
net-bond-reconfigure-delay         default  17
no-proxy                           default  127.0.0.1,localhost,::1
provisioner-harvest-mode           default  destroyed
proxy-ssh                          default  false
resource-tags                      model    {}
snap-http-proxy                    default  ""
snap-https-proxy                   default  ""
snap-store-assertions              default  ""
snap-store-proxy                   default  ""
snap-store-proxy-url               default  ""
ssl-hostname-verification          default  true
storage-default-filesystem-source  model    lxd
test-mode                          default  false
transmit-vendor-metrics            default  true
update-status-hook-interval        default  5m

```

**Comment by SimonRichardson** [comment]
https://trello.com/c/rhsDxjK4/2736-bug-1879972-juju-model-config-does-not-export-in-a-way-that-is-reusable

**Comment by SimonRichardson** [comment]
Closing in favour of #11796
